src/gr-convert-units.c

```diff
@@ -0,0 +1,325 @@
+/* gr-convert-units.c:
+ *
+ * Copyright (C) 2017 Paxana Xander <paxana@paxana.me>
+ *
+ * Licensed under the GNU General Public License Version 3
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+ 
+#include <string.h>
+#include <stdlib.h>
+#include <glib/gi18n.h>
+#include <locale.h>
+#include <langinfo.h>
+#include <gtk/gtk.h>
+#include <math.h>
+#include "config.h"
+#include "gr-settings.h"
+#include "gr-unit.h"
+#include "gr-convert-units.h"
+#include "gr-number.h"
+#include "gr-utils.h"
+ 
+GrTemperatureUnit
+gr_convert_get_temperature_unit (void)
+{
+        int unit;
+        GSettings *settings = gr_settings_get ();
+
+        unit = g_settings_get_enum (settings, "temperature-unit");
+
+        if (unit == GR_TEMPERATURE_UNIT_LOCALE) {
+        #ifdef LC_MEASUREMENT
+            const char *fmt;
+
+            fmt = nl_langinfo (_NL_MEASUREMENT_MEASUREMENT);
+        
+            if (fmt && *fmt == 2)
+                unit = GR_TEMPERATURE_UNIT_FAHRENHEIT;
+            else
+        #endif
+                unit = GR_TEMPERATURE_UNIT_CELSIUS;
+        }
+
+        return unit;
+}
+ 
+GrPreferredUnit
+gr_convert_get_volume_unit (void)
+ {
+    int unit;
+    GSettings *settings = gr_settings_get ();
+
+    unit = g_settings_get_enum (settings, "volume-unit");
+
+    if (unit == GR_PREFERRED_UNIT_LOCALE) {
+#ifdef LC_MEASUREMENT
+                const char *fmt;
+
+            fmt = nl_langinfo (_NL_MEASUREMENT_MEASUREMENT);
+            if (fmt && *fmt == 2)
+                unit = GR_PREFERRED_UNIT_IMPERIAL;
+            else
+#endif
+                unit = GR_PREFERRED_UNIT_METRIC;
+        }
+
+        return unit;
+}
+ 
+GrPreferredUnit
+gr_convert_get_weight_unit (void)
+{
+    int unit;
+    GSettings *settings = gr_settings_get ();
+ 
+    unit = g_settings_get_enum (settings, "weight-unit");
+ 
+        if (unit == GR_PREFERRED_UNIT_LOCALE) {
+ 
+            #ifdef LC_MEASUREMENT
+                 const char *fmt;
+ 
+                fmt = nl_langinfo (_NL_MEASUREMENT_MEASUREMENT);
+                
+                if (fmt && *fmt == 2)
+                    unit = GR_PREFERRED_UNIT_IMPERIAL;
+                else
+            #endif
+                    unit = GR_PREFERRED_UNIT_METRIC;
+        }
+        return unit;
+}
```

Comment:
Looking at these 3 functions, it strikes me that this is essentially the same code. The only difference is the setting name. Maybe we should turn this into a little helper function that takes the setting as an extra argument: 

GrPreferredUnit
gr_convert_get_metric_or_imperial (const char *setting)
{
...
}

and then

GrPreferredUnit
gr_convert_weight_unit (void)
{
  return gr_convert_get_metric_or_imperial ("weight-unit");
}